compiler/bindings/python/iree/compiler/tools/import_onnx/__main__.py

```diff
@@ -58,13 +61,64 @@ def main(args):
         print(m.get_asm(assume_verified=not args.no_verify))
 
 
-def load_onnx_model(file_path: Path) -> onnx.ModelProto:
-    raw_model = onnx.load(file_path)
-    inferred_model = onnx.shape_inference.infer_shapes(raw_model)
+def load_onnx_model(args: argparse.Namespace) -> onnx.ModelProto:
+    # Do shape inference two ways.  First, attempt in-memory to avoid redundant
+    # loading and the need for writing a temporary file somewhere.  If that
+    # fails, typically because of the 2 GB protobuf size limit, try again via
+    # files.  See
+    # https://github.com/onnx/onnx/blob/main/docs/PythonAPIOverview.md#shape-inference-a-large-onnx-model-2gb
+    # for details about the file-based technique.
+
+    # Make a temp dir for all the temp files we'll be generating as a side
+    # effect of infering shapes.  For now, the only file is a new .onnx holding
+    # the revised model with shapes.
+    #
+    # TODO: If the program temp_dir is None, we should be using an ephemeral
+    # temp directory instead of a hard-coded path in order to avoid data races
+    # by default.
+    input_dir = os.path.dirname(os.path.abspath(args.input_file))
+    temp_dir = tempfile.TemporaryDirectory(dir=input_dir)
+    temp_dir_path = Path(temp_dir.name)
+
+    # Load the model, with possible external data coming from the default
+    # location, or the location specified on the command line.
+    if args.data_dir is None:
+        raw_model = onnx.load(args.input_file)
+    else:
+        raw_model = onnx.load(args.input_file, load_external_data=False)
+        onnx.load_external_data_for_model(raw_model, args.data_dir)
```

Comment:
Reorder this - postpone creating the temporary directory until it is actually needed. If the model fails to load here, we don't want to end up with orphaned files. Using a context manager for the files would also help (vs the explicit `.cleanup()`.

```python
def load_onnx_model(args):
  model = onnx.load(...)

  # Perform shape inference, first attempting it in-memory then falling back to using files.
  try:
    inferred_model = onnx.shape_inference...
    return inferred_model
  except ValueError as e:
    # (I'm not sure about the syntax here, should test it)
    logger.info("In-memory shape inference failed:", e)

  with tempfile.TemporaryDirectory() as tmpdirname:
    onnx.shape_inference.infer_shapes_path(...)
    ...
    return inferred_model
```

The program should follow a logical ordering. As written now, there are several flags that interact with one another and only sometimes actually affect behavior.

Reply:
I reordered, but with tempfile, whenever the program completes (crash or ends), the temporary files and directories are automatically deleted.

Reply:
This doesn't look fully reordered yet.

* Load
* Try shape inference in-memory
* Create temp dir
* Try shape inference using a file

The "# Do shape inference two ways." comment block should be next to where shape inference is actually happening (that could also be a helper function, but the script is fairly small so NBD)

Reply:
Ok moved the comments around, but I think the actual code is following this flow